### ISSUE TITLE
MassFormer

### ISSUE THREAD

**Issue by yunyicheng** [huggingface/transformers#25293]
### Model description

We propose adding a new model, MassFormer, to predict tandem mass spectra accurately. MassFormer uses a graph transformer architecture to model long-distance relationships between atoms in the molecule. The transformer module is initialized with parameters obtained through a chemical pre-training task, then fine-tuned on spectral data. MassFormer outperforms competing approaches for spectrum prediction on multiple datasets and is able to recover prior knowledge about the effect of collision energy on the spectrum. We demonstrate that the model can identify relationships between fragment peaks by employing gradient-based attribution methods. To further highlight MassFormer’s utility, we show that it can match or exceed existing prediction-based methods on two spectrum identification tasks. Our code is the first open-source implementation of a deep-learning MS/MS spectrum predictor and may encourage future research in this area.

### Open source status

- [X] The model implementation is available
- [X] The model weights are available

### Provide useful links for the implementation

This model will be implemented according to the paper by @adamoyoung as listed below.

Reference:
Young, A., Wang, B. and Röst, H., 2021. MassFormer: Tandem mass spectrum prediction with graph transformers. arXiv preprint arXiv:2111.04824.

**Comment by kamathis4** [comment]
Hey @amyeroberts, could I implement this?
I think the implemented link is [here](https://github.com/Roestlab/massformer)

**Comment by yunyicheng** [comment]
Hi Adithya, thank you for your interest in this project! However, this is my undergraduate research project in Rost Lab, and I am working with Adamo Young now, planning to optimize the original implementation and make it more suitable for hugging face.

**Comment by kamathis4** [comment]
Oh that's alright @yunyicheng, good luck on the implementation!!

**Comment by amyeroberts** [comment]
Hi @yunyicheng, thanks for opening this new model feature request! 

The recommended and easiest way to add a model is adding the model code to the hub directly using [this tutorial](https://huggingface.co/docs/transformers/custom_models). This avoids the lengthy review process and your model is still available through the typical transformers API e.g. `AutoModel.from_pretrained(...)`. 

cc @clefourrier for graph models :)

**Comment by clefourrier** [comment]
Hi! I agree that adding the code to the hub directly would be the best way to go :)
